File: tests/docker-compat/python-client/tests/keys.py
"""Category 7: Key Management tests."""

import time as time_mod
from .framework import (
    register, run_test, key, assert_equal, assert_true,
    assert_equal_int,
)


def test_keys(r, prefix):
    results = []
    cat = "keys"

    def t_del():
        k1 = key(prefix, "key:del1")
        k2 = key(prefix, "key:del2")
        k3 = key(prefix, "key:del3")
        r.set(k1, "a")
        r.set(k2, "b")
        r.set(k3, "c")
        n = r.delete(k1)
        assert_equal_int(1, n)
        n = r.delete(k2, k3)
        assert_equal_int(2, n)
    results.append(run_test("DEL_single_and_multiple", cat, t_del))

    def t_exists():
        k1 = key(prefix, "key:ex1")
        k2 = key(prefix, "key:ex2")
        try:
            r.set(k1, "a")
            r.set(k2, "b")
            n = r.exists(k1)
            assert_equal_int(1, n)
            n = r.exists(k1, k2, key(prefix, "key:nonexist"))
            assert_equal_int(2, n)
        finally:
            r.delete(k1, k2)
    results.append(run_test("EXISTS_single_and_multiple", cat, t_exists))

    def t_expire_ttl():
        k = key(prefix, "key:expire")
        try:
            r.set(k, "val")
            ok = r.expire(k, 10)
            assert_true(ok, "EXPIRE should succeed")
            ttl = r.ttl(k)
            assert_true(0 < ttl <= 10, "TTL should be between 0 and 10s")
        finally:
            r.delete(k)
    results.append(run_test("EXPIRE_TTL", cat, t_expire_ttl))

    def t_pexpire_pttl():
        k = key(prefix, "key:pexpire")
        try:
            r.set(k, "val")
            ok = r.pexpire(k, 10000)
            assert_true(ok, "PEXPIRE should succeed")
            pttl = r.pttl(k)
            assert_true(pttl > 0, "PTTL should be positive")
        finally:
            r.delete(k)
    results.append(run_test("PEXPIRE_PTTL", cat, t_pexpire_pttl))

    def t_expireat():
        k = key(prefix, "key:expireat")
        try:
            r.set(k, "val")
            future = int(time_mod.time()) + 60
            ok = r.expireat(k, future)
            assert_true(ok, "EXPIREAT should succeed")
            ttl = r.ttl(k)
            assert_true(ttl > 50, "TTL should be ~60s")

            k2 = key(prefix, "key:pexpireat")
            r.set(k2, "val")
            ok = r.pexpireat(k2, future * 1000)
            assert_true(ok, "PEXPIREAT should succeed")
        finally:
            r.delete(k, key(prefix, "key:pexpireat"))
    results.append(run_test("EXPIREAT_PEXPIREAT", cat, t_expireat))

    def t_persist():
        k = key(prefix, "key:persist")
        try:
            r.set(k, "val", ex=10)
            ok = r.persist(k)
            assert_true(ok, "PERSIST should succeed")
            ttl = r.ttl(k)
            # -1 means no TTL
            assert_true(ttl == -1, "TTL should be -1 after PERSIST (no expiration)")
        finally:
            r.delete(k)
    results.append(run_test("PERSIST", cat, t_persist))

    def t_type():
        ks = key(prefix, "key:type_s")
        kl = key(prefix, "key:type_l")
        kh = key(prefix, "key:type_h")
        kz = key(prefix, "key:type_z")
        kset = key(prefix, "key:type_set")
        try:
            r.set(ks, "val")
            r.lpush(kl, "val")
            r.hset(kh, "f", "v")
            r.zadd(kz, {"a": 1})
            r.sadd(kset, "a")
            assert_equal("string", r.type(ks))
            assert_equal("list", r.type(kl))
            assert_equal("hash", r.type(kh))
            assert_equal("zset", r.type(kz))
            assert_equal("set", r.type(kset))
        finally:
            r.delete(ks, kl, kh, kz, kset)
    results.append(run_test("TYPE", cat, t_type))

    def t_rename():
        k1 = key(prefix, "key:rename_src")
        k2 = key(prefix, "key:rename_dst")
        k3 = key(prefix, "key:renamenx_dst")
        try:
            r.set(k1, "val")
            r.rename(k1, k2)
            assert_equal("val", r.get(k2))
            # RENAMENX
            r.set(k1, "new")
            r.set(k3, "exists")
            ok = r.renamenx(k1, k3)
            assert_true(not ok, "RENAMENX should fail when dest exists")
            r.delete(k3)
            r.set(k1, "new")
            ok = r.renamenx(k1, k3)
            assert_true(ok, "RENAMENX should succeed when dest doesn't exist")
        finally:
            r.delete(k1, k2, k3)
    results.append(run_test("RENAME_RENAMENX", cat, t_rename))

    def t_keys_pattern():
        k1 = key(prefix, "key:pat_a")
        k2 = key(prefix, "key:pat_b")
        k3 = key(prefix, "key:pat_c")
        try:
            r.set(k1, "1")
            r.set(k2, "2")
            r.set(k3, "3")
            keys = r.keys(key(prefix, "key:pat_*"))
            assert_true(len(keys) >= 3, "KEYS should return at least 3 keys")
        finally:
            r.delete(k1, k2, k3)
    results.append(run_test("KEYS_pattern", cat, t_keys_pattern))

    def t_scan():
        test_keys = []
        for i in range(10):
            k = key(prefix, f"key:scan_{i}")
            test_keys.append(k)
            r.set(k, str(i))
        try:
            all_keys = []
            cursor = 0
            while True:
                cursor, keys = r.scan(cursor, match=key(prefix, "key:scan_*"), count=5)
                all_keys.extend(keys)
                if cursor == 0:
                    break
            assert_true(len(all_keys) >= 10, f"SCAN should find at least 10 keys, got {len(all_keys)}")
        finally:
            for k in test_keys:
                r.delete(k)
    results.append(run_test("SCAN_with_pattern", cat, t_scan))

    def t_unlink():
        k = key(prefix, "key:unlink")
        r.set(k, "val")
        n = r.unlink(k)
        assert_equal_int(1, n)
        assert_equal_int(0, r.exists(k))
    results.append(run_test("UNLINK", cat, t_unlink))

    def t_object_encoding():
        k = key(prefix, "key:obj_enc")
        try:
            r.set(k, "12345")
            enc = r.object("encoding", k)
            assert_true(len(str(enc)) > 0, "OBJECT ENCODING should return a value")
        finally:
            r.delete(k)
    results.append(run_test("OBJECT_ENCODING", cat, t_object_encoding))

    def t_copy():
        src = key(prefix, "key:copy_src")
        dst = key(prefix, "key:copy_dst")
        try:
            r.set(src, "copyval")
            res = r.copy(src, dst)
            assert_true(res, "COPY should succeed")
            assert_equal("copyval", r.get(dst))
        finally:
            r.delete(src, dst)
    results.append(run_test("COPY", cat, t_copy))

    def t_dump_restore():
        src = key(prefix, "key:dump")
        dst = key(prefix, "key:restore")
        try:
            # Need a non-decode_responses client for DUMP/RESTORE
            import redis as redis_lib
            rb = redis_lib.Redis(
                host=r.connection_pool.connection_kwargs["host"],
                port=r.connection_pool.connection_kwargs["port"],
                password=r.connection_pool.connection_kwargs.get("password"),
                db=0, decode_responses=False,
            )
            rb.set(src, "dumpval")
            dump = rb.dump(src)
            assert_true(len(dump) > 0, "DUMP should return data")
            rb.restore(dst, 0, dump)
            val = rb.get(dst)
            assert_equal(b"dumpval", val)
            rb.close()
        finally:
            r.delete(src, dst)
    results.append(run_test("DUMP_RESTORE", cat, t_dump_restore))

    return results


register("keys", test_keys)

File: tests/docker-compat/js-client/tests/keys.js
const { register, runTest, key, assertEqual, assertTrue, assertEqualInt } = require('./framework');

register('keys', async (redis, prefix) => {
  const results = [];
  const cat = 'keys';

  results.push(await runTest('DEL_single_and_multiple', cat, async () => {
    const k1 = key(prefix, 'key:del1');
    const k2 = key(prefix, 'key:del2');
    const k3 = key(prefix, 'key:del3');
    await redis.set(k1, 'a');
    await redis.set(k2, 'b');
    await redis.set(k3, 'c');
    const n1 = await redis.del(k1);
    assertEqualInt(1, n1);
    const n2 = await redis.del(k2, k3);
    assertEqualInt(2, n2);
  }));

  results.push(await runTest('EXISTS_single_and_multiple', cat, async () => {
    const k1 = key(prefix, 'key:ex1');
    const k2 = key(prefix, 'key:ex2');
    await redis.set(k1, 'a');
    await redis.set(k2, 'b');
    const n1 = await redis.exists(k1);
    assertEqualInt(1, n1);
    const n2 = await redis.exists(k1, k2, key(prefix, 'key:nonexist'));
    assertEqualInt(2, n2);
    await redis.del(k1, k2);
  }));

  results.push(await runTest('EXPIRE_TTL', cat, async () => {
    const k = key(prefix, 'key:expire');
    await redis.set(k, 'val');
    const ok = await redis.expire(k, 10);
    assertEqualInt(1, ok);
    const ttl = await redis.ttl(k);
    assertTrue(ttl > 0 && ttl <= 10, 'TTL should be between 0 and 10');
    await redis.del(k);
  }));

  results.push(await runTest('PEXPIRE_PTTL', cat, async () => {
    const k = key(prefix, 'key:pexpire');
    await redis.set(k, 'val');
    const ok = await redis.pexpire(k, 10000);
    assertEqualInt(1, ok);
    const pttl = await redis.pttl(k);
    assertTrue(pttl > 0, 'PTTL should be positive');
    await redis.del(k);
  }));

  results.push(await runTest('EXPIREAT_PEXPIREAT', cat, async () => {
    const k = key(prefix, 'key:expireat');
    await redis.set(k, 'val');
    const future = Math.floor(Date.now() / 1000) + 60;
    const ok1 = await redis.expireat(k, future);
    assertEqualInt(1, ok1);
    const ttl = await redis.ttl(k);
    assertTrue(ttl > 50, 'TTL should be ~60s');

    const k2 = key(prefix, 'key:pexpireat');
    await redis.set(k2, 'val');
    const futureMs = Date.now() + 60000;
    const ok2 = await redis.pexpireat(k2, futureMs);
    assertEqualInt(1, ok2);
    await redis.del(k, k2);
  }));

  results.push(await runTest('PERSIST', cat, async () => {
    const k = key(prefix, 'key:persist');
    await redis.set(k, 'val', 'EX', 10);
    const ok = await redis.persist(k);
    assertEqualInt(1, ok);
    const ttl = await redis.ttl(k);
    // -1 means no expiry, -2 means key doesn't exist
    assertTrue(ttl < 0, 'TTL should be negative after PERSIST (no expiration)');
    await redis.del(k);
  }));

  results.push(await runTest('TYPE', cat, async () => {
    const ks = key(prefix, 'key:type_s');
    const kl = key(prefix, 'key:type_l');
    const kh = key(prefix, 'key:type_h');
    const kz = key(prefix, 'key:type_z');
    const kset = key(prefix, 'key:type_set');
    await redis.set(ks, 'val');
    await redis.lpush(kl, 'val');
    await redis.hset(kh, 'f', 'v');
    await redis.zadd(kz, 1, 'a');
    await redis.sadd(kset, 'a');

    assertEqual('string', await redis.type(ks));
    assertEqual('list', await redis.type(kl));
    assertEqual('hash', await redis.type(kh));
    assertEqual('zset', await redis.type(kz));
    assertEqual('set', await redis.type(kset));

    await redis.del(ks, kl, kh, kz, kset);
  }));

  results.push(await runTest('RENAME_RENAMENX', cat, async () => {
    const k1 = key(prefix, 'key:rename_src');
    const k2 = key(prefix, 'key:rename_dst');
    const k3 = key(prefix, 'key:renamenx_dst');
    await redis.set(k1, 'val');
    await redis.rename(k1, k2);
    const val = await redis.get(k2);
    assertEqual('val', val);

    // RENAMENX
    await redis.set(k1, 'new');
    await redis.set(k3, 'exists');
    const ok1 = await redis.renamenx(k1, k3);
    assertEqualInt(0, ok1);
    await redis.del(k3);
    const ok2 = await redis.renamenx(k1, k3);
    assertEqualInt(1, ok2);

    await redis.del(k1, k2, k3);
  }));

  results.push(await runTest('KEYS_pattern', cat, async () => {
    const k1 = key(prefix, 'key:pat_a');
    const k2 = key(prefix, 'key:pat_b');
    const k3 = key(prefix, 'key:pat_c');
    await redis.set(k1, '1');
    await redis.set(k2, '2');
    await redis.set(k3, '3');
    const keys = await redis.keys(key(prefix, 'key:pat_*'));
    assertTrue(keys.length >= 3, 'KEYS should return at least 3 keys');
    await redis.del(k1, k2, k3);
  }));

  results.push(await runTest('SCAN_with_pattern', cat, async () => {
    for (let i = 0; i < 10; i++) {
      await redis.set(key(prefix, `key:scan_${i}`), String(i));
    }
    const all = [];
    let cursor = '0';
    do {
      const [nextCursor, keys] = await redis.scan(cursor, 'MATCH', key(prefix, 'key:scan_*'), 'COUNT', 5);
      cursor = nextCursor;
      all.push(...keys);
    } while (cursor !== '0');
    assertTrue(all.length >= 10, `SCAN should find at least 10 keys, got ${all.length}`);
    for (let i = 0; i < 10; i++) {
      await redis.del(key(prefix, `key:scan_${i}`));
    }
  }));

  results.push(await runTest('UNLINK', cat, async () => {
    const k = key(prefix, 'key:unlink');
    await redis.set(k, 'val');
    const n = await redis.unlink(k);
    assertEqualInt(1, n);
    const exists = await redis.exists(k);
    assertEqualInt(0, exists);
  }));

  results.push(await runTest('OBJECT_ENCODING', cat, async () => {
    const k = key(prefix, 'key:obj_enc');
    await redis.set(k, '12345');
    const enc = await redis.object('ENCODING', k);
    assertTrue(enc !== null && enc.length > 0, 'OBJECT ENCODING should return a value');
    await redis.del(k);
  }));

  results.push(await runTest('COPY', cat, async () => {
    const src = key(prefix, 'key:copy_src');
    const dst = key(prefix, 'key:copy_dst');
    await redis.set(src, 'copyval');
    const res = await redis.call('COPY', src, dst);
    assertEqualInt(1, res);
    const val = await redis.get(dst);
    assertEqual('copyval', val);
    await redis.del(src, dst);
  }));

  results.push(await runTest('DUMP_RESTORE', cat, async () => {
    const src = key(prefix, 'key:dump');
    const dst = key(prefix, 'key:restore');
    await redis.set(src, 'dumpval');
    const dump = await redis.dumpBuffer(src);
    assertTrue(dump !== null && dump.length > 0, 'DUMP should return data');
    await redis.restore(dst, 0, dump);
    const val = await redis.get(dst);
    assertEqual('dumpval', val);
    await redis.del(src, dst);
  }));

  return results;
});
